vernac/vernacexpr.ml

```diff
@@ -286,6 +288,22 @@ type extend_name =
 
 type discharge = DoDischarge | NoDischarge
 
+type hint_info_expr = Constrexpr.constr_pattern_expr Typeclasses.hint_info_gen
+
+type reference_or_constr =
+  | HintsReference of Libnames.qualid
+  | HintsConstr of Constrexpr.constr_expr
+
+type hints_expr =
+  | HintsResolve of (hint_info_expr * bool * reference_or_constr) list
+  | HintsResolveIFF of bool * Libnames.qualid list * int option
+  | HintsImmediate of reference_or_constr list
+  | HintsUnfold of Libnames.qualid list
+  | HintsTransparency of Libnames.qualid Hints.hints_transparency_target * bool
+  | HintsMode of Libnames.qualid * Hints.hint_mode list
+  | HintsConstructors of Libnames.qualid list
+  | HintsExtern of int * Constrexpr.constr_expr option * Genarg.raw_generic_argument
```

Comment:
Why did this stuff move? Maybe copy the explanation to the commit message.

Reply:
Actually I had to scratch my head quite a bit with this; the dependency cycle was:
- `Declare: Vernacexpr.decl_notation list`
- `Vernacexpr: ComHint.hint_expr`
- `ComHint`: `Declare.declare_constant`

There are quite a few ways to solve this, for example by adding the AST to `Metasyntax` , having `Declare` not to depend on `Vernacexpr` [which seems actually sane in the medium term], reworking notation support.... I chose moving hints ast to vernacexpr as the approach least invasive and more consistent with the current style, but indeed we may want implement some other solution in the medium term.

Comment added to the commit message.